core/src/main/java/com/opensymphony/xwork2/ognl/OgnlValueStack.java

```diff
@@ -61,6 +62,12 @@
 
     private static final String MAP_IDENTIFIER_KEY = "com.opensymphony.xwork2.util.OgnlValueStack.MAP_IDENTIFIER_KEY";
 
+    private static final String EXCEPTION_EVALUATING_EXPRESSION_STR = "Caught an exception while evaluating expression '{}' against value stack";
+
+    private static final String DEVMODE_SAFETY_STR = "Setting development mode [{}] affects the safety of your application!";
+
+    private static final AtomicBoolean INITIAL_WARNING_FOR_DEV_MODE = new AtomicBoolean(true);
```

Comment:
Thank you @JCgH4164838Gh792C124B5 but still it seems too complicated logic for me for just a log line. I think we can revert it to what it was before. The phrase `devMode` I think already conventionally WARNs user it's for development environment.

And could you please revise all other logs introduced in same PR's commits if they may raise similar issues? Thanks in advance!

Reply:
Hello Again.

During the review of logs introduced in the previous PR, I decided to go ahead and apply the same only-warn-once-logic to the relevant modules (OgnlUtil, OgnlValueStackFactory, CompoundRootAccessor, DefaultAcceptedPatternsChecker, DefaultExcludedPatternsChecker).  Also simplified the `devMode` log warn for OgnlValueStack, and committed the end-result for consideration.

If the reasoning in the previous comment is considered sufficient (as it applies to all of the preceding), then some or all of the changes could be applied ... and if nothing else it makes the the logging locations are easier to track for this PR.  :)  :)

Reply:
Thank you @JCgH4164838Gh792C124B5 for your suggested benefits for Struts :) yes I already knew and understand what are you're trying to achieve. Actually I not only was worry for performance but also for somehow more important thing, the code readability and nice logic/design. Benefit#1 is good but I think the injection point isn't good place to log because of three reasons: 1) In future, anyone who creates a new class that needs devMode has to remember or simply will forget to consider this log consideration 2) It violates DRY (Don't Repeat Yourself) 3) For not singleton beans, it causes log floods and an angry user :) I guess not only your new added log statements but also already existent e.g. [here](https://github.com/apache/struts/pull/291/files#diff-ac4c516bdbcef4755bd30af8d088e5b5L51) may suffer from this issue. It depends, it may work on singleton beans but still suffers from reason#1&2. Benefit#2 already has been addressed as much as was logically possible by refactoring public to protected. Any further improvements are welcomed only if they don't violate reason#1&2 or any other best practice :)

I think reviewing #291 and reverting potential issues for now. Then we can consider these benefits in a more appropriate place (rather than a bottle neck like prototype beans injection points). wdyt?

Reply:
Hello @yasserzamani .
Thanks for the review.  :)

I just wanted to make sure of my points before going ahead and reverting most of the log entries introduced with #291 .  Thanks for taking the time to provide the counterpoints to the ones I raised and giving good reasons as to why the added logging is more-trouble-than-its-worth.  :) :)

The last commit has been amended again and effectively removes all new logging introduced (except for the accepted/excluded locations, that if I understood correctly, @lukaszlenart indicated should probably remain).  If I misunderstood that point it's a quick fix.

Reply:
Thanks again! I reviewed `AcceptedPatternsChecker` and `ExcludedPatternsChecker` interfaces and how they work. I feel their new warn logs should be removed or changed to debug level as well. Because firstly `this affects all instances and safety of your application` seems isn't quite true. Those interfaces are `prototype` beans injected to parameters and cookie interceptors. Maybe user adds an accepted pattern to only one specific action. This doesn't affects all instances and safety of application but user sees numerous of this warning; each action call prints this :(

Reply:
@yasserzamani those are prototypes on purpose - interceptors are singletons per package which means those instance will be injected just once into the interceptors.